src/scan.ts

```
@@ -0,0 +1,99 @@
+
+import ERR, { Result, siftValues } from './err'
+import { ensureDocumentOf } from './linkage'
+import { report } from './schema'
+import { Maybe } from './is'
+
+export const ErrRepeated = ERR `ErrRepeated` ((props: { fqname: string }) =>
+  `non-repeatable directive "${props.fqname}" was found multiple times`)
+
+export const ErrRepetition = ERR `ErrRepetition` ((props: { fqname: string }) =>
+  `"${props.fqname}" found here`)
+
+export const scan = <D extends AnyDirective>(
+  node: Directive_NodeOf<D>,
+  directive: D
+): Directive_DataOf<D>[] =>
+  scannerFor(directive)(node)
+
+const scannerFor = derive (
+  'scanner', <D extends AnyDirective>(directive: D): (node: Directive_NodeOf<D>) => Directive_DataOf<D>[] => {        
+    const { spec, name, args, repeatable, locations } = directive
+    const self = Spec.parse(spec)
+    const fqname = `${spec}#${name}`
+    const kinds = new Set(locations.map(locToKind))
+    
+    type Args = Directive_ArgsOf<D>
+    const forDoc = derive(
+      `impl ${fqname}`,
+      (doc): Maybe<Dir<Args>> => {
+        const specifiedName = namespaceFor(doc, self)?.specifiedName(name)
+        if (!specifiedName) return null
+        return dir(specifiedName, args)
+      })
+
+    return derive(
+      `results ${fqname}`, (node: Directive_NodeOf<D>): Directive_DataOf<D>[] => {
+        if (!kinds.has(node.kind)) return []
+        const de = forDoc(ensureDocumentOf(node))
+
+        if (!de) return []
+        const results: Result<Shape_DeTypeOf<Args>>[] =
+          directivesOn(node).map(de.deserialize)
+            // Include errors, exclude nulls
+            .filter(r => r.is === 'err' || r.ok) as Result<Shape_DeTypeOf<Args>>[]
+        if (!repeatable && results.length > 1) {
+          report(ErrRepeated({ fqname, node }),
+            ...results.map(r => ErrRepetition({ fqname, node: r.node })))
+          return []
+        }
+        const [errors, values] = siftValues(results)
+        report(...errors)        
```

I don't really understand error handling in this function.  If the error is "multiple directives on non-repeatable" then we stash that error on the doc and return empty list to the caller.  But if the error is "we couldn't parse something" then we stash the error on the doc and we still return data to the caller.  And then this function only has one caller (in isExport) which just looks at the return value without doing anything.  It seems weird to me that that caller should be just carrying on with its business even if what it saw was `core__surface(export: 123)`? I guess eventually the user is responsible for calling `ensure` but it feels like failing faster makes sense here?